kani-compiler/src/kani_middle/transform/check_uninit/ty_layout.rs

```diff
@@ -0,0 +1,275 @@
+pub enum PointeeLayout {
+    Static { layout: TypeLayout },
+    Slice { element_layout: TypeLayout },
+    TraitObject,
+}
+
+pub struct PointeeInfo {
+    pointee_ty: Ty,
+    layout: PointeeLayout,
+}
+
+impl PointeeInfo {
+    pub fn from_ty(ty: Ty) -> Result<Self, String> {
+        match ty.kind() {
+            TyKind::RigidTy(rigid_ty) => match rigid_ty {
+                RigidTy::Str => {
+                    let slicee_ty = Ty::unsigned_ty(UintTy::U8);
+                    let size_in_bytes = slicee_ty.layout().unwrap().shape().size.bytes();
+                    let data_chunks = data_bytes_for_ty(&MachineInfo::target(), slicee_ty, 0)?;
+                    let layout = PointeeLayout::Slice {
+                        element_layout: TypeLayout { size_in_bytes, data_chunks },
+                    };
+                    Ok(PointeeInfo { pointee_ty: ty, layout })
+                }
+                RigidTy::Slice(slicee_ty) => {
+                    let size_in_bytes = slicee_ty.layout().unwrap().shape().size.bytes();
+                    let data_chunks = data_bytes_for_ty(&MachineInfo::target(), slicee_ty, 0)?;
+                    let layout = PointeeLayout::Slice {
+                        element_layout: TypeLayout { size_in_bytes, data_chunks },
+                    };
+                    Ok(PointeeInfo { pointee_ty: ty, layout })
+                }
+                RigidTy::Dynamic(..) => {
+                    Ok(PointeeInfo { pointee_ty: ty, layout: PointeeLayout::TraitObject })
+                }
+                _ => {
+                    if ty.layout().unwrap().shape().is_sized() {
+                        let size_in_bytes = ty.layout().unwrap().shape().size.bytes();
+                        let data_chunks = data_bytes_for_ty(&MachineInfo::target(), ty, 0)?;
+                        let layout = PointeeLayout::Static {
+                            layout: TypeLayout { size_in_bytes, data_chunks },
+                        };
+                        Ok(PointeeInfo { pointee_ty: ty, layout })
+                    } else {
+                        Err(
+                            "Can only determine unsized type layout information for slices and trait objects.".to_string(),
+                        )
+                    }
+                }
+            },
+            TyKind::Alias(..) | TyKind::Param(..) | TyKind::Bound(..) => {
+                Err("Can only determine type layout information for RigidTy.".to_string())
+            }
+        }
+    }
+
+    pub fn ty(&self) -> &Ty {
+        &self.pointee_ty
+    }
+
+    pub fn layout(&self) -> &PointeeLayout {
+        &self.layout
+    }
+}
+
+/// Get a size of an initialized scalar.
+fn scalar_ty_size(machine_info: &MachineInfo, ty: Ty) -> Option<DataBytes> {
+    let shape = ty.layout().unwrap().shape();
+    match shape.abi {
+        ValueAbi::Scalar(Scalar::Initialized { value, .. })
+        | ValueAbi::ScalarPair(Scalar::Initialized { value, .. }, _) => {
```

Comment:
Don't you need the second scalar in this case?

Reply:
Talked about it offline. Now it uses both of the scalars for fat pointers.